// checkly/dist/commands/deploy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const api = require("../rest/api");
const config_1 = require("../services/config");
const inquirer_1 = require("inquirer");
const core_1 = require("@oclif/core");
const authCommand_1 = require("./authCommand");
const project_parser_1 = require("../services/project-parser");
const checkly_config_loader_1 = require("../services/checkly-config-loader");
const api_1 = require("../rest/api");
const constructs_1 = require("../constructs");
const chalk = require("chalk");
const check_1 = require("../constructs/check");
const alert_channel_1 = require("../constructs/alert-channel");
const util_1 = require("../services/util");
const common_messages_1 = require("../messages/common-messages");
// eslint-disable-next-line no-restricted-syntax
var ResourceDeployStatus;
(function (ResourceDeployStatus) {
    ResourceDeployStatus["UPDATE"] = "UPDATE";
    ResourceDeployStatus["CREATE"] = "CREATE";
    ResourceDeployStatus["DELETE"] = "DELETE";
})(ResourceDeployStatus || (ResourceDeployStatus = {}));
class Deploy extends authCommand_1.AuthCommand {
    async run() {
        var _a, _b, _c, _d, _e, _f, _g;
        core_1.ux.action.start('Parsing your project', undefined, { stdout: true });
        const { flags } = await this.parse(Deploy);
        const { force, preview, output, config: configFilename } = flags;
        const { configDirectory, configFilenames } = (0, util_1.splitConfigFilePath)(configFilename);
        const { config: checklyConfig, constructs: checklyConfigConstructs, } = await (0, checkly_config_loader_1.loadChecklyConfig)(configDirectory, configFilenames);
        const { data: avilableRuntimes } = await api_1.runtimes.getAll();
        const project = await (0, project_parser_1.parseProject)({
            directory: configDirectory,
            projectLogicalId: checklyConfig.logicalId,
            projectName: checklyConfig.projectName,
            repoUrl: checklyConfig.repoUrl,
            checkMatch: (_a = checklyConfig.checks) === null || _a === void 0 ? void 0 : _a.checkMatch,
            browserCheckMatch: (_c = (_b = checklyConfig.checks) === null || _b === void 0 ? void 0 : _b.browserChecks) === null || _c === void 0 ? void 0 : _c.testMatch,
            ignoreDirectoriesMatch: (_d = checklyConfig.checks) === null || _d === void 0 ? void 0 : _d.ignoreDirectoriesMatch,
            checkDefaults: checklyConfig.checks,
            browserCheckDefaults: (_e = checklyConfig.checks) === null || _e === void 0 ? void 0 : _e.browserChecks,
            availableRuntimes: avilableRuntimes.reduce((acc, runtime) => {
                acc[runtime.name] = runtime;
                return acc;
            }, {}),
            checklyConfigConstructs,
        });
        core_1.ux.action.stop();
        const { data: account } = await api.accounts.get(config_1.default.getAccountId());
        if (!force && !preview) {
            const { confirm } = await (0, inquirer_1.prompt)([{
                    name: 'confirm',
                    type: 'confirm',
                    message: `You are about to deploy your project "${project.name}" to account "${account.name}". Do you want to continue?`,
                }]);
            if (!confirm) {
                return;
            }
        }
        try {
            const projectPayload = project.synthesize(false);
            const { data } = await api.projects.deploy(projectPayload, { dryRun: preview });
            if (preview || output) {
                this.log(this.formatPreview(data, project));
            }
            if (!preview) {
                await core_1.ux.wait(500);
                this.log(`Successfully deployed project "${project.name}" to account "${account.name}".`);
            }
        }
        catch (err) {
            if (((_f = err === null || err === void 0 ? void 0 : err.response) === null || _f === void 0 ? void 0 : _f.status) === 400) {
                throw new Error(`Failed to deploy your project due to a missing field. ${(_g = err.response.data) === null || _g === void 0 ? void 0 : _g.message}`);
            }
            else {
                throw new Error(`Failed to deploy your project. ${err.message}`);
            }
        }
    }
    formatPreview(previewData, project) {
        var _a, _b;
        // Current format of the data is: { checks: { logical-id-1: 'UPDATE' }, groups: { another-logical-id: 'CREATE' } }
        // We convert it into update: [{ logicalId, resourceType, construct }, ...], create: [], delete: []
        // This makes it easier to display.
        const updating = [];
        const creating = [];
        const deleting = [];
        for (const change of (_a = previewData === null || previewData === void 0 ? void 0 : previewData.diff) !== null && _a !== void 0 ? _a : []) {
            const { type, logicalId, action } = change;
            if (type === constructs_1.AlertChannelSubscription.__checklyType) {
                // Don't report changes to alert channel subscriptions.
                // User's don't create these directly, so it's more intuitive to consider it as part of the check.
                continue;
            }
            const construct = project.data[type][logicalId];
            if (action === ResourceDeployStatus.UPDATE) {
                updating.push({ resourceType: type, logicalId, construct });
            }
            else if (action === ResourceDeployStatus.CREATE) {
                creating.push({ resourceType: type, logicalId, construct });
            }
            else if (action === ResourceDeployStatus.DELETE) {
                // Since the resource is being deleted, the construct isn't in the project.
                deleting.push({ resourceType: type, logicalId });
            }
        }
        if (!creating.length && !deleting.length && !updating.length) {
            return '\nNo checks were detected. More information on how to set up a Checkly CLI project is available at https://checklyhq.com/docs/cli/.\n';
        }
        // Having some order will make the output easier to read.
        const compareEntries = (a, b) => a.resourceType.localeCompare(b.resourceType) ||
            a.logicalId.localeCompare(b.logicalId);
        updating.sort(compareEntries);
        creating.sort(compareEntries);
        deleting.sort(compareEntries);
        const output = [];
        if (creating.length) {
            output.push(chalk.bold.green('Create:'));
            for (const { logicalId, construct } of creating) {
                output.push(`    ${construct.constructor.name}: ${logicalId}`);
            }
            output.push('');
        }
        if (deleting.length) {
            output.push(chalk.bold.red('Delete:'));
            const prettyResourceTypes = {
                [check_1.Check.__checklyType]: 'Check',
                [alert_channel_1.AlertChannel.__checklyType]: 'AlertChannel',
                [constructs_1.CheckGroup.__checklyType]: 'CheckGroup',
            };
            for (const { resourceType, logicalId } of deleting) {
                output.push(`    ${(_b = prettyResourceTypes[resourceType]) !== null && _b !== void 0 ? _b : resourceType}: ${logicalId}`);
            }
            output.push('');
        }
        if (updating.length) {
            output.push(chalk.bold.magenta('Update and Unchanged:'));
            for (const { logicalId, construct } of updating) {
                output.push(`    ${construct.constructor.name}: ${logicalId}`);
            }
            output.push('');
        }
        return output.join('\n');
    }
}
exports.default = Deploy;
Deploy.hidden = false;
Deploy.description = 'Deploy your project to your Checkly account.';
Deploy.flags = {
    preview: core_1.Flags.boolean({
        char: 'p',
        description: 'Show a preview of the changes made by the deploy command.',
        default: false,
    }),
    output: core_1.Flags.boolean({
        char: 'o',
        description: 'Shows the changes made after the deploy command.',
        default: false,
    }),
    force: core_1.Flags.boolean({
        char: 'f',
        description: common_messages_1.default.forceMode,
        default: false,
    }),
    config: core_1.Flags.string({
        char: 'c',
        description: common_messages_1.default.configFile,
    }),
};
