// checkly/dist/commands/destroy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const api = require("../rest/api");
const checkly_config_loader_1 = require("../services/checkly-config-loader");
const authCommand_1 = require("./authCommand");
const inquirer_1 = require("inquirer");
const config_1 = require("../services/config");
const util_1 = require("../services/util");
const common_messages_1 = require("../messages/common-messages");
class Destroy extends authCommand_1.AuthCommand {
    async run() {
        const { flags } = await this.parse(Destroy);
        const { force, config: configFilename } = flags;
        const { configDirectory, configFilenames } = (0, util_1.splitConfigFilePath)(configFilename);
        const { config: checklyConfig } = await (0, checkly_config_loader_1.loadChecklyConfig)(configDirectory, configFilenames);
        const { data: account } = await api.accounts.get(config_1.default.getAccountId());
        if (!force) {
            const { projectName } = await (0, inquirer_1.prompt)([{
                    name: 'projectName',
                    type: 'test',
                    message: `Are you sure you want to delete all resources in project "${checklyConfig.projectName}" for account "${account.name}"?\n  Please confirm by typing the project name "${checklyConfig.projectName}":`,
                }]);
            if (projectName !== checklyConfig.projectName) {
                this.log(`The entered project name "${projectName}" doesn't match the expected project name "${checklyConfig.projectName}".`);
                return;
            }
        }
        await api.projects.deleteProject(checklyConfig.logicalId);
        this.log(`All resources associated with project "${checklyConfig.projectName}" have been successfully deleted.`);
    }
}
exports.default = Destroy;
Destroy.hidden = false;
Destroy.description = 'Destroy your project with all its related resources.';
Destroy.flags = {
    force: core_1.Flags.boolean({
        char: 'f',
        description: common_messages_1.default.forceMode,
        default: false,
    }),
    config: core_1.Flags.string({
        char: 'c',
        description: common_messages_1.default.configFile,
    }),
};
