// checkly/dist/services/config.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
const conf_1 = require("conf");
const dataSchema = {
    accountId: { type: 'string' },
    accountName: { type: 'string' },
};
const authSchema = {
    apiKey: { type: 'string' },
};
const projectSuffix = (_a = process.env.CHECKLY_ENV) !== null && _a !== void 0 ? _a : '';
// eslint-disable-next-line no-restricted-syntax
var Env;
(function (Env) {
    Env["production"] = "production";
    Env["staging"] = "staging";
    Env["development"] = "development";
    Env["test"] = "test";
})(Env || (Env = {}));
const config = {
    auth: new conf_1.default({
        projectName: '@checkly/cli',
        configName: 'auth',
        projectSuffix,
        // @ts-ignore
        schema: authSchema,
    }),
    data: new conf_1.default({
        projectName: '@checkly/cli',
        configName: 'config',
        projectSuffix,
        // @ts-ignore
        schema: dataSchema,
    }),
    clear() {
        this.auth.clear();
        this.data.clear();
    },
    getEnv() {
        const environments = ['production', 'staging', 'development', 'test'];
        const env = process.env.CHECKLY_ENV || environments[0];
        if (!(env in Env)) {
            throw new Error('Invalid CHECKLY_ENV');
        }
        return env;
    },
    getApiKey() {
        return process.env.CHECKLY_API_KEY || this.auth.get('apiKey') || '';
    },
    getAccountId() {
        return process.env.CHECKLY_ACCOUNT_ID || this.data.get('accountId') || '';
    },
    hasEnvVarsConfigured() {
        const apiKey = process.env.CHECKLY_API_KEY || '';
        const accoundId = process.env.CHECKLY_ACCOUNT_ID || '';
        return apiKey !== '' || accoundId !== '';
    },
    getApiUrl() {
        const environments = {
            development: 'http://localhost:3000',
            test: 'https://api-test.checklyhq.com',
            staging: 'https://api-test.checklyhq.com',
            production: 'https://api.checklyhq.com',
        };
        return environments[this.getEnv()];
    },
    getMqttUrl() {
        const environments = {
            development: 'wss://events-local.checklyhq.com',
            test: 'wss://events-test.checklyhq.com',
            staging: 'wss://events-test.checklyhq.com',
            production: 'wss://events.checklyhq.com',
        };
        return environments[this.getEnv()];
    },
    hasValidCredentials() {
        if (this.getApiKey() !== '' && this.getAccountId() !== '') {
            return true;
        }
        return false;
    },
};
exports.default = config;
